lambdas/es/indexer/lam.py

```diff
@@ -63,33 +61,46 @@ def transform_meta(meta):
     }
     return result
 
-def format_codecell(cell):
-    # outputs: [{output_type: stream, text: [str]}]
-    # OR [{output_type: execute_result, data: {'text/plain': [str]}}]
-    #
-    # source: [str]
-    formatted_source = ' '.join(cell['source'])
-    formatted_output = ''
-    if cell['outputs']:
-        first_output = cell['outputs'][0]
-        if first_output['output_type'] == 'stream':
-            formatted_output = ' '.join(first_output['text'])
-        if (first_output['output_type'] == 'data'
-                and first_output['data'].get('text/plain', None)):
-            formatted_output = ' '.join(first_output['data']['text/plain'])
-
-    return formatted_source + formatted_output
-
-def format_notebook(nb):
-    cells = nb['cells']
-    codecells = filter(lambda x: x['cell_type'] == 'code', cells)
-    formatted = map(format_codecell, codecells)
-    text = ' '.join(list(formatted))
-    markdowncells = filter(lambda x: x['cell_type'] == 'markdown', cells)
-    markdowntext = ' '.join([' '.join(cell['source']) for cell in markdowncells])
-    return text + ' ' + markdowntext
+def extract_text(notebook_str):
+    """ Extract code and markdown
+    Args:
+        * nb - notebook as a string
+    Returns:
+        * str - select code and markdown source (and outputs)
+    Pre:
+        * notebook is well-formed per notebook version 4
+        * 'cell_type' is defined for all cells
+        * 'source' defined for all 'code' and 'markdown' cells
+    Throws:
+        * Anything nbformat.reads() can throw :( which is diverse and poorly
+        documented, hence the `except Exception` in handler()
+    Notes:
+        * Deliberately decided not to index output streams and display strings
+        because they were noisy and low value
+        * Tested this code against ~6400 Jupyter notebooks in
+        https://alpha.quiltdata.com/b/alpha-quilt-storage/tree/notebook-search/
+        * Might be useful to index "cell_type" : "raw" in the future
+    See also:
+        * Format reference https://nbformat.readthedocs.io/en/latest/format_description.html
+    """
+    formatted = nbformat.reads(notebook_str, as_version=NB_VERSION)
+    code = []
+    markdown = []
+    for cell in formatted.get('cells', []):
+        if 'source' in cell and 'cell_type' in cell:
+            if cell['cell_type'] == 'code':
+                code.append(cell['source'])
+            elif cell['cell_type'] == 'markdown':
+                markdown.append(cell['source'])
+
+    return '\n'.join(code + markdown)
```

Comment:
This does not keep the code and markdown in order -- it will do all the code, then all the markdown. Just use one output list, and put all code and markdown in it as you go through.

Reply:
it's now in-order